File: generate_amp.py
import feedparser
import os
import re
from datetime import datetime

# LINK FEED REALE DI BLOGGER
FEED_URL = "https://www.brunorachiele.it/feeds/posts/default?alt=rss"

# CARTELLA AMP
AMP_DIR = "amp"

# IMMAGINE FISSA PER TUTTI GLI ARTICOLI
FIXED_IMAGE = "https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEib-mwyxO9DqQx2o08DZF34q2BfiC3YeW9NebmMSRWxZ92ApQBwbBhuwhQhU9UeKYDDa_IqpLTDDHm9Em931n5KSDEjbMlJM4nmXhDgpplo_m6gZTR-3S8KOmWPNfzSZ5c0dlxcm0REM0RkMMGsF-BVpzsSQO0gmdrv6pD6n6kaXOYid7LpzueECp1sQ10/s1600/POLITICA.png"

# NUMERO MASSIMO DI ARTICOLI
MAX_ARTICLES = 10

# CREA CARTELLA SE NON ESISTE
os.makedirs(AMP_DIR, exist_ok=True)

feed = feedparser.parse(FEED_URL)

# funzione per creare slug
def slugify(text):
    text = text.lower()
    text = re.sub(r"[^\w\s-]", "", text)
    text = re.sub(r"\s+", "-", text)
    return text.strip("-")

# Genera pagine AMP
for entry in feed.entries[:MAX_ARTICLES]:
    title = entry.title
    link = entry.link
    pubDate = datetime(*entry.published_parsed[:6]).strftime("%Y-%m-%d")
    slug = slugify(title)
    file_path = f"{AMP_DIR}/{slug}.html"

    amp_html = f"""<!doctype html>
<html ⚡ lang="it">
<head>
  <meta charset="utf-8">
  <title>{title} | Bruno Rachiele</title>
  <link rel="canonical" href="{link}">
  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
  <meta name="description" content="{title}">
  <meta name="robots" content="index, follow">

  <script async src="https://cdn.ampproject.org/v0.js"></script>
  <script async custom-element="amp-ad" src="https://cdn.ampproject.org/v0/amp-ad-0.1.js"></script>

  <style amp-boilerplate>
body{{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;
-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;
-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;
animation:-amp-start 8s steps(1,end) 0s 1 normal both}}
@-webkit-keyframes -amp-start{{from{{visibility:hidden}}to{{visibility:visible}}}}
@-moz-keyframes -amp-start{{from{{visibility:hidden}}to{{visibility:visible}}}}
@-ms-keyframes -amp-start{{from{{visibility:hidden}}to{{visibility:visible}}}}
@-o-keyframes -amp-start{{from{{visibility:hidden}}to{{visibility:visible}}}}
@keyframes -amp-start{{from{{visibility:hidden}}to{{visibility:visible}}}}
</style>
<noscript>
<style amp-boilerplate>
body{{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}}
</style>
</noscript>

  <style amp-custom>
    body{{font-family:Arial,sans-serif;max-width:760px;margin:auto;padding:16px;color:#111;}}
    h1{{font-size:1.7em;margin:15px 0;}}
    .date{{font-size:0.85em;color:#666;margin-bottom:10px;}}
    amp-img{{border-radius:8px;margin:15px 0;width:100%;height:auto;}}
    .cta{{display:block;text-align:center;margin:25px 0;padding:14px;background:#6598b7;color:#fff;font-size:1.1em;border-radius:6px;text-decoration:none;}}
    .ad{{margin:25px 0;}}
  </style>
</head>
<body>

<h1>{title}</h1>
<p class="date">{pubDate}</p>

<amp-img src="{FIXED_IMAGE}" width="1200" height="675" layout="responsive" alt="{title}"></amp-img>

<a class="cta" href="{link}">👉 Leggi l’articolo completo su brunorachiele.it</a>

<div class="ad">
  <amp-ad width="100vw" height="320"
    type="adsense"
    data-ad-client="ca-pub-9225028785900171"
    data-ad-slot="2980836148"
    data-auto-format="rspv"
    data-full-width="">
    <div overflow=""></div>
  </amp-ad>
</div>

</body>
</html>"""

    with open(file_path, "w", encoding="utf-8") as f:
        f.write(amp_html)

print(f"Generate {min(MAX_ARTICLES,len(feed.entries))} AMP articles in '{AMP_DIR}/'")

File: scripts/generate-amp.js
const Parser = require('rss-parser');
const fs = require('fs-extra');
const slugify = require('slugify');

const parser = new Parser();

async function generateAMP() {
  try {
    console.log("🚀 Avvio generazione pagine AMP...");

    // Feed Blogger corretto
    const feedURL = 'https://www.brunorachiele.it/feeds/posts/default?alt=rss';
    console.log("📡 Scarico feed RSS da:", feedURL);

    const feed = await parser.parseURL(feedURL);

    if (!feed.items || feed.items.length === 0) {
      console.warn("⚠️ Nessun articolo trovato nel feed RSS!");
      return;
    }

    console.log(`✅ Trovati ${feed.items.length} articoli`);

    // Assicurati che la cartella amp esista
    fs.ensureDirSync('amp');

    let indexItems = [];

    for (let i = 0; i < feed.items.length; i++) {
      const post = feed.items[i];
      const slug = slugify(post.title || `articolo-${i+1}`, { lower: true, strict: true });

      // Estrai prima immagine grande dal contenuto HTML
      let thumbnail = null;
      const htmlContent = post['content:encoded'] || '';
      const imgMatches = [...htmlContent.matchAll(/<img[^>]+src="([^">]+)"/gi)];
      if (imgMatches.length > 0) thumbnail = imgMatches[0][1];

      const snippet = post.contentSnippet
        ? post.contentSnippet.replace(/<[^>]*>?/gm, '').substring(0, 200) + '...'
        : '';

      indexItems.push({
        title: post.title,
        link: `amp/${slug}.html`,
        pubDate: post.pubDate || '',
        contentSnippet: snippet,
        thumbnail: thumbnail
      });

      // Genera pagina singola AMP
      const articleHtml = `<!doctype html>
<html ⚡ lang="it">
<head>
<meta charset="utf-8">
<title>${post.title} | Bruno Rachiele</title>
<link rel="canonical" href="${post.link}">
<meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
<meta name="description" content="${snippet}">
<meta name="robots" content="index, follow">

<script async src="https://cdn.ampproject.org/v0.js"></script>

<style amp-boilerplate>
body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;
      -moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;
      -ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;
      animation:-amp-start 8s steps(1,end) 0s 1 normal both}
@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}
@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}
@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}
@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}
@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}
</style>
<noscript>
<style amp-boilerplate>
body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}
</style>
</noscript>

<style amp-custom>
body { font-family: Arial,sans-serif; max-width:800px; margin:auto; padding:20px; color:#111; }
.logo { display:block; margin:0 auto 20px; width:120px; height:120px; }
h1 { text-align:center; margin-bottom:20px; }
p { line-height:1.6; margin:4px 0; }
amp-img { margin-bottom:10px; }
a.read-more { display:inline-block; margin-top:10px; font-weight:bold; color:#1a73e8; }
</style>

</head>
<body>
<img class="logo" src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEiOeKLAFzNySkI5ahC_uZiv0Asne6shnjr1QndoMNmRBtkNdQfe59GGO3nY5nc5C4OCXNMXGXu1XtHm0X7gPqbvTgFjqszFGQrPD3G5R4c2MIJPiZCk5OX6dLbnYKeU1itEVo7OcJB9I0N-Wi9z9WP64l5_btoNMrZ111Ex8i-W9UcKwOJPv5uDnl6LwhA/s1600/2ed26cc2-38e4-4d67-a1c5-5e522b654c90%20%281%29.png" alt="Logo Bruno Rachiele">
<h1>${post.title}</h1>
${thumbnail ? `<amp-img src="${thumbnail}" width="600" height="400" layout="responsive" alt="${post.title}"></amp-img>` : ''}
<p>${snippet}</p>
<a class="read-more" href="${post.link}" target="_blank">Leggi tutto sul sito</a>
</body>
</html>`;

      fs.writeFileSync(`amp/${slug}.html`, articleHtml.trim());
      console.log(`📝 Pagina AMP generata: amp/${slug}.html`);
    }

    // Genera feed.json con ultimi 10 articoli
    fs.writeFileSync('feed.json', JSON.stringify({ items: indexItems.slice(0, 10) }, null, 2));
    console.log("✅ feed.json aggiornato");

    // Genera index.html AMP
    const indexHtml = `<!doctype html>
<html ⚡ lang="it">
<head>
<meta charset="utf-8">
<title>Ultime Notizie | Bruno Rachiele</title>
<link rel="canonical" href="https://amp.brunorachiele.it/">
<meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
<meta name="description" content="Ultime notizie aggiornate automaticamente dal sito Bruno Rachiele.">
<meta name="robots" content="index, follow">

<script async src="https://cdn.ampproject.org/v0.js"></script>
<script async custom-element="amp-list" src="https://cdn.ampproject.org/v0/amp-list-0.1.js"></script>
<script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>

<style amp-boilerplate>
body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;
      -moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;
      -ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;
      animation:-amp-start 8s steps(1,end) 0s 1 normal both}
@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}
@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}
@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}
@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}
@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}
</style>
<noscript>
<style amp-boilerplate>
body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}
</style>
</noscript>

<style amp-custom>
body { font-family: Arial,sans-serif; max-width:800px; margin:auto; padding:20px; color:#111; }
.logo { display:block; margin:0 auto 20px; width:120px; height:120px; }
h1 { text-align:center; margin-bottom:20px; }
article { margin-bottom:30px; border-bottom:1px solid #ddd; padding-bottom:15px; }
h2 { font-size:1.4em; margin:0 0 8px; }
a { color:#1a73e8; text-decoration:none; }
p { line-height:1.6; margin:4px 0; }
amp-img { margin-bottom:10px; }
.read-more { display:inline-block; margin-top:8px; font-weight:bold; color:#1a73e8; }
</style>
</head>
<body>
<img class="logo" src="logo.png" alt="Logo Bruno Rachiele">
<h1>Ultime Notizie</h1>
<amp-list width="auto" height="1600" layout="fixed-height" src="feed.json" items="items">
<template type="amp-mustache">
<article>
{{#thumbnail}}
<amp-img src="{{thumbnail}}" width="600" height="400" layout="responsive" alt="{{title}}"></amp-img>
{{/thumbnail}}
<h2><a href="{{link}}">{{title}}</a></h2>
<p>{{pubDate}}</p>
<p>{{contentSnippet}}</p>
<a class="read-more" href="{{link}}">Leggi tutto</a>
</article>
</template>
</amp-list>
</body>
</html>`;

    fs.writeFileSync('index.html', indexHtml.trim());
    console.log("✅ index.html AMP generato");

    console.log("🎉 Generazione completata con successo!");
  } catch (err) {
    console.error("❌ Errore durante la generazione AMP:", err.message);
  }
}

generateAMP();
